```{r setup, echo=FALSE, include=FALSE, cache=FALSE}
# Set up the environment
library(knitr) 
opts_chunk$set(cache.path='cache/', fig.path='img/', cache=T, tidy=T, fig.keep='high', echo=T, dpi=300, out.width=700, echo=F)
options(replace.assign=TRUE, width=120)
set.seed(1)
```

```{r loadLibraries, warning=FALSE, message=FALSE}
library(biomaRt)
# Use the latest biomart version for mm
mart <- useMart("ensembl", dataset="mmusculus_gene_ensembl")
source("work_scripts//venn5.R")
library(gridExtra)
library(DESeq2)
annotDb <- "org.Mm.eg.db"
library(package=annotDb, character.only=T)
library(GO.db)
library(GOstats)
library(reactome.db)
# Preparing environment for remapping Gene Symbols to Entrez IDs
x <- org.Mm.egSYMBOL2EG
# Get the entrez gene identifiers that are mapped to a gene symbol
mapped_genes <- mappedkeys(x)
# Convert to a list
xx <- as.list(x[mapped_genes])
```

```{r Enrichment, echo=FALSE}
# Perform GO enrichment analysis of topTable produced by limma.
# use use for KEGG of GO enrichment analyses
# Provide which ontology space to use, "MF", "BP", of "CC"
# Returns a list of the results, and total count of significant enrichments
Enrichment <- function(res, use="GO", fileName=NULL, ont="BP")
{
  res.genes <- res
  # Convert selected and all gene names to Entrez IDs, removing NAs
  sel.entrez <- unlist(xx)[res.genes]; sel.entrez <- sel.entrez[!is.na(sel.entrez)]
  all.entrez <- unlist(xx) # All Entrez genes
  # Prepare parameters for the enrichment analysis
  if (use == "GO")
    {
    params <- new('GOHyperGParams', geneIds=sel.entrez, universeGeneIds=all.entrez, ontology=ont,
 pvalueCutoff=0.05, conditional=F, testDirection='over', annotation=annotDb)
    }
 else
   {
    params <- new('KEGGHyperGParams', geneIds=sel.entrez, universeGeneIds=all.entrez, pvalueCutoff=0.05, testDirection='over', annotation=annotDb) 
   }
  hgOver <- hyperGTest(params)
  result <- summary(hgOver)
  result <- cbind(result, p.adjust(result$Pvalue, method="BH")) # Append corrected for multiple testing p-value
  colnames(result)[length(colnames(result))] <- "p.adj"
  result <- result[result$p.adj < 0.1, ] # Subset the results keeping FDR at 10%
  if (!is.null(fileName)) {
    write.table(result, paste("results//", fileName, sep=""), sep="\t", row.names=F)
  }
  return(list(result, nrow(result)))
}
```

```{r GOEnrichment}
# Perform GO enrichment analysis of topTable produced by limma.
# Provide which ontology space to use, "MF", "BP", of "CC"
# Returns a list of the results, and total count of significant enrichments
GOEnrichment <- function(res, ont="BP", fileName=NULL)
{
  # Convert selected and all gene names to Entrez IDs, removing NAs
  sel.entrez <- unlist(xx)[res]; sel.entrez <- sel.entrez[!is.na(sel.entrez)]
  all.entrez <- unlist(xx)# All Entrez genes
  # Prepare parameters for the GO enrichment analysis
  params <- new('GOHyperGParams',
                geneIds=sel.entrez,
                universeGeneIds=all.entrez,
                ontology=ont,
                pvalueCutoff=0.05,
                conditional=F,
                testDirection='over',
                annotation="org.Mm.eg.db"
  )
  hgOver <- hyperGTest(params)
  result <- summary(hgOver)
  result <- cbind(result, p.adjust(result$Pvalue, method="BH")) # Append corrected for multiple testing p-value
  colnames(result)[length(colnames(result))] <- "p.adj"
  result <- result[result$p.adj < 0.1, ] # Subset the results keeping FDR at 10%
  if (!is.null(fileName)) {
    write.table(result, paste("results//", fileName, sep=""), sep="\t", row.names=F)
  }
  return(list(result, nrow(result)))
}
```

```{r reactomeEnrichment, echo=FALSE}
# Perform Reactome pathway enrichment analysis of topTable produced by limma
# Returns a list of the results, and total count of significant enrichments
reactomeEnrichment <- function(res, fileName=NULL)
{
  # Convert selected and all gene names to Entrez IDs, removing NAs
  sel.entrez <- unlist(xx)[res]; sel.entrez <- sel.entrez[!is.na(sel.entrez)]
  all.entrez <- unlist(xx); all.entrez <- all.entrez[!is.na(all.entrez)]
  # Prepare pathways and counts
  entrez2path <- as.list(reactomeEXTID2PATHID) # EntrezIDs to pathway IDs mapping
  sel.counts <- table(unlist(entrez2path[sel.entrez])) # Selected gene counts in pathways
  all.counts <- table(unlist(entrez2path[all.entrez])) # All gene counts
  n.sel <- length(sel.entrez) # Total number of selected genes
  n.all <- length(all.entrez) # Total number of all genes
  pathways <- names(sel.counts) # PAthways to test for enrichment
  pathway.names <- as.list(reactomePATHID2NAME) # Pathway ID to name mapping
  # For each pathway, construct 2x2 contingency table and calculate the enrichment p-value using Fisher's exact test
  result <- lapply(pathways, function(p) 
  {
                   sel <- sel.counts[[p]] 
                   all <- all.counts[[p]]
                   pathway.name <- pathway.names[[p]]
                   m <- matrix(c(sel, all, n.sel-sel, n.all-all),
                               byrow=T,nrow=2)
                   p.value <- fisher.test(m, alternative="greater")$p.value
                   data.frame(PathwayID=p, 
                              Pathway=pathway.name,
                              N.Selected=sel,
                              N.In.Category=all,
                              p.value=p.value)
  }
  ) # End lapply
  result <- as.data.frame(do.call(rbind, result)) # Convert list to data frame
  result <- result[order(result$p.value),] # Order by p-value
  result <- cbind(result, p.adjust(result$p.value, method="BH")) # Append corrected for multiple testing p-value
  colnames(result)[length(colnames(result))] <- "p.adj"
  # result <- result[result$p.adj < 0.1, ] # Subset the results keeping FDR at 10%
  if (!is.null(fileName)) {
    write.table(result, paste("results//", fileName, sep=""), sep="\t", row.names=F)
  }
  return(list(result, sum(result$p.adj < 0.05)))
}
```

```{r GO}
# Plot top 20 enriched GOs or KEGGs
GOKEGG <- function(x, use="GO", filename=NULL) {
  res.go <- Enrichment(x[, 1], use, fileName=filename)
  message("The number of enriched ", use, " :", res.go[[2]])
  if (res.go[[2]] > 20) {i <- 20} else {i <- res.go[[2]]}
  if (i > 0) {grid.table(res.go[[1]][1:i, ], gp=gpar(fontsize=7))}
}
```

```{r Pathway}
# Plot top 20 enriched Reactome
Pathway <- function(x, filename=NULL) {
  res.pathway <- reactomeEnrichment(x[, 1], filename)
  message("The number of enriched Reactome pathways:", res.pathway[[2]])
  if (res.pathway[[2]] > 20) {i <- 20} else {i <- res.pathway[[2]]}
  if (i > 0) {grid.table(res.pathway[[1]][1:i, ], gp=gpar(fontsize=7))}
}
```

```{r biomart}
refseq2gene <- function(x, fileName) {
  # Function to convert NM_ IDs to gene names/descriptions
  genes <- getBM(attributes=c('external_gene_id', 'description'), filters='refseq_mrna', values=x, mart=mart, uniqueRows=T)
  if (!is.null(fileName)) {
    write.table(genes, paste("results//", fileName, sep=""), sep="\t", row.names=F)
  }
  return(genes)
}
ens2gene <- function(x, fileName=NULL) {
  # Function to convert NM_ IDs to gene names/descriptions
  genes <- getBM(attributes=c('external_gene_id', 'description'), filters='ensembl_gene_id', values=x, mart=mart, uniqueRows=T)
  if (!is.null(fileName)) {
    write.table(genes, paste("results//", fileName, sep=""), sep="\t", row.names=F)
  }
  return(genes)
}
gene2ens <- function(x, fileName=NULL) {
  # Function to convertgene names to ensembl IDs
  genes <- getBM(attributes=c('external_gene_id', 'ensembl_gene_id'), filters='wikigene_name', values=x, mart=mart, uniqueRows=T)
  if (!is.null(fileName)) {
    write.table(genes, paste("results//", fileName, sep=""), sep="\t", row.names=F)
  }
  return(genes)
}
```

```{r loadData, results='hide', message=FALSE}
# Load the data ant get DEGs
readDDS <- function(DIR) {
  sampleFiles <- list.files(path=DIR, pattern=".+txt$")
  sampleFiles <- sampleFiles[grep("control4|mutant1", sampleFiles, invert=T)] # remove CTL4 and MUT1 outliers
  sampleCondition <- sub("^control.*", "control", sampleFiles)
  sampleCondition[sampleCondition != "control"] <- "mutant"
  sampleTable <- data.frame(sampleName = sampleFiles,
                            fileName = sampleFiles,
                            condition = sampleCondition)
  ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                         directory = DIR,
                                         design = ~ condition)
  ddsHTSeq$condition <- factor(ddsHTSeq$condition, levels=c("control", "mutant"))
  return(ddsHTSeq)
}

DIR <- "data//htseq-count_rev/"
ddsHTSEQ <- lapply(DIR, readDDS)

# Getting differentially expressed genes
resDDS <- function(DDS, pval=0.01, lfc=0) {
  dds <- DESeq(DDS)
  res <- results(dds)
  res <- res[complete.cases(res), ]
  return(res[res$padj < pval & abs(res$log2FoldChange) > lfc, ])
}
```

Differential gene expression
========================================================
```{r get_degs, echo=FALSE}
res <- lapply(ddsHTSEQ, resDDS)
# Write results to file
write.table(merge(res[[1]], getBM(attributes=c('ensembl_gene_id', 'external_gene_id', 'description'), filters='ensembl_gene_id', values=rownames(res[[1]]), mart=mart, uniqueRows=T), by.x="row.names", by.y=1, all.x=T), paste("results//", "DEGs_all.txt", sep=""), sep="\t", row.names=T)
```

First, we check how many genes are differentially expressed

```{r allGenes, results='hide'}
# Get all genes from sig_genes_names[[1]], no repeatSoaker applied
genes_all_nor <- unique(ens2gene(rownames(res[[1]]), "genes_all_nor.txt")[, 1])
message(paste("Number of differentially expressed PROBES:", nrow(res[[1]])))
message(paste("Number of differentially expressed GENES:", length(genes_all_nor)))
```

Some probes map to multiple genes, hence the discrepancy in counts of probes and genes. 

KEGG enrichment analysis
----------------------------------------------
```{r KEGGall_nor}
GOKEGG(matrix(genes_all_nor, ncol=1), use="KEGG", "KEGG_all_nor.txt")
```

GO enrichment analysis
----------------------------------------------
```{r GOall_nor}
GOKEGG(matrix(genes_all_nor, ncol=1), use="GO", "GO_all_nor.txt")
```

Reactome enrichment analysis
----------------------------------------------
```{r Pathwayall_nor}
Pathway(matrix(genes_all_nor, ncol=1), "PW_all_nor.txt")
```

Plots of selected genes
------------------------
```{r barboxplots}
# Function plotting error bars. http://monkeysuncle.stanford.edu/?p=485
error.bar <- function(x, y, upper, lower=upper, length=0.1,...){
  if(length(x) != length(y) | length(y) !=length(lower) | length(lower) != length(upper))
    stop("vectors must be same length")
  arrows(x,y+upper, x, y-lower, angle=90, code=3, length=length, ...)
}

# Function to extract data subset
data.sel <- function(selectedGenes="", mtx) {
  selectedEns <- gene2ens(selectedGenes)
  selectedData <- mtx[rownames(mtx) %in% selectedEns[, "ensembl_gene_id"],]
  selectedEns[selectedEns[, "ensembl_gene_id"] %in% rownames(selectedData), "external_gene_id"]
  data <- merge(selectedData, selectedEns, by.x="row.names", by.y="ensembl_gene_id")
  rownames(data) <- data[,ncol(data)]
  data <- as.matrix(data[,c(-1, -ncol(data))])
  return(data)
}


# Plot bar/box plot
plot.bb <- function(selectedGenes, data, type="bar") {
# More blunt but working method, processing one row at a time
#dev.off() # Clear the plot area
  par(mfrow=c(2,2)) # Num X Num plot area
  names<-c("control","mutant")
  indices <- lapply(names, function(p) {grep(p, colnames(data))}) # Group specific column indices
  for (i in 1:nrow(data)) {  
    if (type == "bar") {
      data.means<-c(rowMeans(data[i,indices[[1]], drop=F]),rowMeans(data[i,indices[[2]], drop=F]))
      data.sd<-1.96*c(apply(data[i,indices[[1]], drop=F],1,sd)/sqrt(length(data[i,indices[[1]]])),apply(data[i,indices[[2]], drop=F],1,sd)/sqrt(length(data[i,indices[[2]]])))
      data.min<-min(as.numeric(data[i,])) - 0.1*min(as.numeric(data[i,])) # Get the min -10%
      data.max<-max(as.numeric(data[i,])) + 0.1*max(as.numeric(data[i,])) # Get the max +10%
      barx<-barplot(data.means, main=row.names(data)[i], names.arg=names, 
               ylim=c(data.min,data.max), xpd=F, ylab="log2-normalized expression", las=2) # Plot a hits for current data
      error.bar(barx,data.means,data.sd)
    } else {
      boxplot(data.frame(control=data[i, indices[[1]]], mutant=data[i, indices[[2]]]), main=rownames(data)[i], las=2)
    }
  }
}
```

First, we select genes of interest. 

```{r selectGenes}
(selectedGenes <- c("Dlk1", "Zfp423", "Zfp521", "Zfml"))
```

Then, we plot the BARplots of their normalized log2-expression level.

```{r barplots, fig.height=7, fig.width=4}
rlg <- rlog(ddsHTSEQ[[1]])
rlogMat <- assay(rlg)
#vsd <- varianceStabilizingTransformation(ddsHTSEQ[[1]])
#vstMat <- assay(vsd)
plot.bb(selectedGenes, data.sel(selectedGenes, rlogMat), "bar")
```

We plot the same data using BOXplots.

```{r boxplots, fig.height=7, fig.width=4}
plot.bb(selectedGenes, data.sel(selectedGenes, rlogMat), "box")
```

We compare these boxplots with plotting raw unnormalized read counts

```{r boxplotsRaw, fig.height=7, fig.width=4}
mtx <- counts(ddsHTSEQ[[1]])
plot.bb(selectedGenes, data.sel(selectedGenes, mtx), "box")
```
